docs/load_balancer.md

```diff
@@ -0,0 +1,278 @@
+Once each cloud provider sets the Status, we will stop copying from the Spec to the Status.
+
+For v1 final:
+* we should remove publicIPs from the Spec, and require that this information is retrieved through the Status.
+However, we would need to change the openstack code to auto-allocate a VIP.
+* we should remove createExternalLoadBalancer.
+
+## kube-proxy pseudocode
+
+```
+for each service {
+    for each service.port {
+        // listen on cluster
+        listen on <internalMagicIp>:0
+        proxyPort = figureOutTheRandomPortWeJustGotAssigned()
+        openPortal(service.portalIP, port.port, proxyPort)
+        
+
+        if service.spec.visibility == "loadbalancer" || (service.spec.visibility == "" && service.createExternalLoadBalancer=true) {
+          if (port.hostPort != 0) {
+            openLoadBalancerPortals(service, port.port, port.hostPort) // uses cloud-provider to figure out correct iptables
+          }
+        } else if service.spec.visibility == "public" {
+          if (port.hostPort != 0) {
+            openPublicPortal(<publicMagicIp>, port.port, port.hostPort)
+          }
+        } else if service.spec.visibility == "" {
+          for each service.publicIP {  // deprecated
+            openPublicPortal(publicIP, port.port, proxyPort) // uses cloud-provider to decide srcip vs dstip (gclb vs rax)
+          }
+        }
+   }
+}
+```
+
+
+## kube-proxy iptables
+
+As part of the implementation of kube-proxy's internal magic, kube-proxy allocates ports that are used locally by
+iptables.  These ports are not visible other than to kube-proxy & iptables.  We need to be sure that these ports will
+not conflict with the cluster-wide ports.  To do that, we must make sure that kube-proxy differentiates between
+internal & external traffic, and uses these internal ports only with internal traffic.
+
+There are a few possible implementation strategies here:
+
+1. Use `sysctl -w net.ipv4.conf.all.route_localnet=1`, and then bind kube-proxy to 127.0.0.1 for internal traffic.  Requires kernel >= 3.6
+1. Use addrtype with iptables (external: `! --src-type local --dst-type local`, internal: `--src-type local --dst-type local`)
+1. Use the network interface to identify traffic (internal: lo, cbr0; external: eth0)
+1. Use the container CIDR range to identify internal (container) traffic (e.g. 10.244.0.0/16).  Requires fixing that CIDR.
+
+I would like to try those strategies, basically working from top to bottom until we find that one that works.  I tried
+the container CIDR strategy, so that should work.
+
+The sysctl strategy is nice, but does not work on old kernels, and so I propose for those creating a dummy internal address.
+I think the easiest way to do that is to create a bridge and assign it an IP.  This is a hack, but should allow us
+to support old kernels until they are replaced.  We should also be able to put this hack in the Salt scripts.
```

Comment:
Going against this:

a) salt
b) two modes, one is guaranteed to break over time

I'd rather keep one imperfect but well tested mode than one great and one crap mode.